src/main/java/com/google/devtools/build/lib/query2/PostAnalysisQueryEnvironment.java

```diff
@@ -631,4 +635,43 @@ public Iterable<BuildConfiguration> getConfigurations() {
       }
     }
   }
+
+  protected abstract OrderOutput getOrderOutput();
+
+  protected abstract Comparator<T> getFullOrderingComparator();
+
+  public Iterable<T> orderResults(Iterable<T> resultList)
+      throws InterruptedException {
+    OrderOutput orderOutput = this.getOrderOutput();
+
+    if (orderOutput == OrderOutput.NO) {
+      return resultList;
+    }
+
+    Digraph<T> graph = new Digraph<>();
+
+    // First pass: create nodes for dependency Digraph
+    for (T result : resultList) {
+      Node<T> node = graph.createNode(result);
+    }
+
+    // Second pass: create edges between nodes
+    for (T result : resultList) {
+      for (T dep : getFwdDeps(ImmutableList.of(result))) {
+        if (graph.getNodeMaybe(dep) != null) {
+          graph.addEdge(
+            graph.getNode(result),
+            graph.createNode(dep)
+          );
+        }
+      }
+    }
+
+    return Iterables.transform(
+      orderOutput == OrderOutput.DEPS
+        ? graph.getTopologicalOrder()
+        : graph.getTopologicalOrder(this.getFullOrderingComparator()),
+      Node::getLabel
```

Comment:
This causes `auto` to use `full` ordering also for the `proto` formatter, but the option documentation says otherwise. Could you please address this?

Reply:
Good point but to do this I will need to move the sorting from here to the callback, which is where it's done for `query`.

What's the timeline for upgrading to bazel 5? If that will be happening soon I should probably be making these changes there instead of on this branch.

Reply:
> Good point but to do this I will need to move the sorting from here to the callback, which is where it's done for query.

Is this possible or would it require a large refactoring? Alternatively, could you something like
```
  @Override
  public OrderOutput getOrderOutput() {
    return cqueryOptions.outputFormat.endsWith("proto") && cqueryOptions.orderOutput == OrderOutput.AUTO ? OrderOutput.DEPS : cqueryOptions.orderOutput;
  }
```

> What's the timeline for upgrading to bazel 5? If that will be happening soon I should probably be making these changes there instead of on this branch.

Good point, we've just upgraded to bazel 5.3, could you change the base of this pr to `databricks-bazel-5.3.0`?

Reply:
Sent out a new PR: https://github.com/databricks/bazel/pull/11